### PR TITLE
[testing-on-gke part 6.3] Use monitoring api for gce vm

### PR DESCRIPTION
**Pull request by gargnitingoogle**
### Description
This is on top of #2493 and is followed up in # .

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - NA
3. Integration tests - NA

### CODE REVIEW

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils.py [hunk 1]
```diff
@@ -138,5 +141,195 @@ def unix_to_timestamp(unix_timestamp: int) -> str:
...
+  results = client.list_time_series(
+      request={
+          "name": project_name,
+          "filter": (
+              'metric.type = "kubernetes.io/container/memory/used_bytes"'
+              # ' AND metric.memory_type = "non-evictable"' # for some reason,
+              # this throws error, so commented it out.
+              f" AND resource.labels.cluster_name = {cluster_name}"
+              f" AND resource.labels.pod_name = {pod_name}"
+              # f" AND resource.labels.container_name = {container_name}"
```
Comment (gargnitingoogle) [inline perfmetrics/scripts/testing_on_gke/examples/utils/utils.py:217]: Shouldn't container_name here be 'gke-gcsfuse-sidecar' ?

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils.py [hunk 2]
```diff
@@ -138,5 +141,195 @@ def unix_to_timestamp(unix_timestamp: int) -> str:
...
+          )
+      )
+  )
+
+
+def isRelevantMonitoringResult(
+    result,
+    cluster_name: str,
+    pod_name: str,
+    # container_name: str,
```
Comment (gargnitingoogle) [inline perfmetrics/scripts/testing_on_gke/examples/utils/utils.py:162]: Either clean up unused code, or re-enable it.

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils.py [hunk 3]
```diff
@@ -138,5 +141,195 @@ def unix_to_timestamp(unix_timestamp: int) -> str:
...
+    namespace_name: str,
```
Comment (gargnitingoogle) [inline perfmetrics/scripts/testing_on_gke/examples/utils/utils.py:163]: remove unused function arguments.

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils.py [hunk 4]
```diff
@@ -138,5 +141,195 @@ def unix_to_timestamp(unix_timestamp: int) -> str:
...
+  )
+
+  relevant_results = [
+      result
+      for result in results
+      if isRelevantMonitoringResult(
+          result,
+          cluster_name,
+          pod_name,
+          # container_name,
```
Comment (gargnitingoogle) [inline perfmetrics/scripts/testing_on_gke/examples/utils/utils.py:233]: remove or re-enable commented code.

File: perfmetrics/scripts/testing_on_gke/examples/dlio/parse_logs.py [hunk 5]
```diff
@@ -26,7 +26,7 @@
 # local library imports
 sys.path.append("../")
 import dlio_workload
-from utils.utils import get_memory, get_cpu, standard_timestamp, is_mash_installed
+from utils.utils import get_memory, get_cpu, unix_to_timestamp, standard_timestamp, is_mash_installed, get_memory_from_monitoring_api, get_cpu_from_monitoring_api, timestamp_to_epoch
 from utils.parse_logs_common import ensureDir, download_gcs_objects, parseLogParserArguments, SUPPORTED_SCENARIOS
```
Comment (gargnitingoogle) [inline perfmetrics/scripts/testing_on_gke/examples/dlio/parse_logs.py:30]: Rename ensureDir, parseLogParserArguments to use `_` in place of camel-case for consistency.

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils.py (hunk 2 above)
Comment (gargnitingoogle) [reply perfmetrics/scripts/testing_on_gke/examples/utils/utils.py:162]: re-enabled

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils.py (hunk 4 above)
Comment (gargnitingoogle) [reply perfmetrics/scripts/testing_on_gke/examples/utils/utils.py:233]: reenabled

File: perfmetrics/scripts/testing_on_gke/examples/dlio/parse_logs.py (hunk 5 above)
Comment (gargnitingoogle) [reply perfmetrics/scripts/testing_on_gke/examples/dlio/parse_logs.py:30]: will do in #2496

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            start=self.start,
+            end=self.end,
+        )
+    )
+
+  def test_timestamp_to_epoch(self):
+    timestamp = '2024-08-21T19:20:25'
+    expected_epoch = 1724268025
+    self.assertEqual(timestamp_to_epoch(timestamp), expected_epoch)
+    pass
```
Comment (kislaykishore) [inline perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py:81]: pass is redundant
```suggestion
```

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+            pod_name=self.pod_name,
+            start=self.start,
+            end=self.end,
+        )
+    )
+
+  def test_timestamp_to_epoch(self):
+    timestamp = '2024-08-21T19:20:25'
+    expected_epoch = 1724268025
+    self.assertEqual(timestamp_to_epoch(timestamp), expected_epoch)
```
Comment (kislaykishore) [inline perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py:80]: Nit:
```suggestion
    self.assertEqual(timestamp_to_epoch('2024-08-21T19:20:25'), 1724268025)
```

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    print(
+        get_memory(
+            project_number=self.project_number,
+            pod_name=self.pod_name,
+            start=self.start,
+            end=self.end,
+        )
+    )
+
+  def test_get_cpu_methods(self):
```
Comment (kislaykishore) [inline perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py:57]: What's the objective of this test and the one above? Is it to verify that the calls don't fail?

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils.py [hunk 9]
```diff
@@ -136,5 +143,190 @@ def unix_to_timestamp(unix_timestamp: int) -> str:
...
+  return int(
+      time.mktime(
+          time.strptime(
+              timestamp.split(".")[0].replace("T", " "), "%Y-%m-%d %H:%M:%S"
+          )
+      )
+  )
+
+
+def isRelevantMonitoringResult(
```
Comment (kislaykishore) [inline perfmetrics/scripts/testing_on_gke/examples/utils/utils.py:160]: Should this be renamed to is_relevant_monitoring_result in line with other functions?

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py (hunk 6 above)
Comment (gargnitingoogle) [reply perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py:81]: done. removed.

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py (hunk 7 above)
Comment (gargnitingoogle) [reply perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py:80]: done

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py (hunk 8 above)
Comment (gargnitingoogle) [reply perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py:57]: >  Is it to verify that the calls don't fail?

yes.

Comment (kislaykishore) [reply perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py:57]: Why invoke `print`?

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils.py (hunk 9 above)
Comment (gargnitingoogle) [reply perfmetrics/scripts/testing_on_gke/examples/utils/utils.py:160]: yes, I'll change it to `_is_relevant_monitoring_result` (_ prefix to indicate that this method is private to this file).

File: perfmetrics/scripts/testing_on_gke/examples/dlio/parse_logs.py [hunk 10]
```diff
@@ -185,7 +193,26 @@ def fetch_cpu_memory_data():
...
+              )
+              r["lowest_cpu"], r["highest_cpu"] = get_cpu_from_monitoring_api(
+                  pod_name=r["pod_name"],
+                  start_epoch=r["start_epoch"],
+                  end_epoch=r["end_epoch"],
+                  project_id=args.project_id,
+                  cluster_name=args.cluster_name,
+                  namespace_name=args.namespace_name,
+              )
+          pass
```
Comment (kislaykishore) [inline perfmetrics/scripts/testing_on_gke/examples/dlio/parse_logs.py:215]: pass is only required if the class definition/method block is empty. Here it's a no-op.

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils.py [hunk 11]
```diff
@@ -136,5 +143,190 @@ def unix_to_timestamp(unix_timestamp: int) -> str:
...
+
+
+def timestamp_to_epoch(timestamp: str) -> int:
+  return int(
+      time.mktime(
+          time.strptime(
+              timestamp.split(".")[0].replace("T", " "), "%Y-%m-%d %H:%M:%S"
+          )
+      )
+  )
```
Comment (kislaykishore) [inline perfmetrics/scripts/testing_on_gke/examples/utils/utils.py:157]: See if this works.
```suggestion
  return int(datetime.strptime('2024-08-21T19:20:25.444', "%Y-%m-%dT%H:%M:%S.%f").timestamp())
```

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    print(
+        get_cpu(
+            project_number=self.project_number,
+            pod_name=self.pod_name,
+            start=self.start,
+            end=self.end,
+        )
+    )
+
+  def test_timestamp_to_epoch(self):
```
Comment (kislaykishore) [inline perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py:77]: Add another test where milliseconds part is non-zero.

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py (hunk 6 above)
Comment (gargnitingoogle) [reply perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py]: pass is added by my vim python plugin by default every time I add a new function using `def` keyword, so it keeps adding them and I keep forgetting to remove them. Removed all the useless `pass` statements now.

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils.py (hunk 9 above)
Comment (gargnitingoogle) [reply perfmetrics/scripts/testing_on_gke/examples/utils/utils.py]: done

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils.py (hunk 11 above)
Comment (gargnitingoogle) [reply perfmetrics/scripts/testing_on_gke/examples/utils/utils.py]: yes it works, with a little modification.

File: perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py (hunk 12 above)
Comment (gargnitingoogle) [reply perfmetrics/scripts/testing_on_gke/examples/utils/utils_test.py:83]: added

File: perfmetrics/scripts/testing_on_gke/examples/dlio/parse_logs.py (hunk 10 above)
Comment (gargnitingoogle) [reply perfmetrics/scripts/testing_on_gke/examples/dlio/parse_logs.py]: done. removed all useless `pass` statements from everywhere.

**Review by kislaykishore — APPROVED** [review]
